File: final_juspay_solutions/reachAbility.java
import java.util.Scanner;

public class reachAbility {

    public static boolean isReachableNode(int nodes, int[] edges, int src, int des, boolean[] vis) {

        vis[src] = true;

        if (edges[src] != -1) {
            int adj = edges[src];

            if (!vis[adj]) {
                vis[adj] = true;
                if (adj == des)
                    return true;
                if (isReachableNode(nodes, edges, adj, des, vis) == true)
                    return true;
            }
        }
        return false;
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int nodes = sc.nextInt();

        int edges[] = new int[nodes];
        for (int i = 0; i < nodes; i++)
            edges[i] = sc.nextInt();

        int src = sc.nextInt();
        int des = sc.nextInt();

        boolean vis[] = new boolean[nodes];

        boolean res = isReachableNode(nodes, edges, src, des, vis);

        System.out.println("The given two nodes are Reachable or Not -->" + res);

        sc.close();
    }
}

File: Latest_Cpp/Reachability.cpp
#include <iostream>
#include <vector>
using namespace std;

bool isReachableNode(int nodes, const vector<int>& edges, int src, int des, vector<bool>& vis) {
    if (src == des) return true;

    vis[src] = true;

    if (edges[src] != -1) {
        int adj = edges[src];

        // Safeguard against invalid indices
        if (adj >= 0 && adj < nodes && !vis[adj]) {
            if (adj == des) return true;
            if (isReachableNode(nodes, edges, adj, des, vis))
                return true;
        }
    }

    return false;
}

int main() {
    int nodes;
    cin >> nodes;

    vector<int> edges(nodes);
    for (int i = 0; i < nodes; ++i)
        cin >> edges[i];

    int src, des;
    cin >> src >> des;

    vector<bool> vis(nodes, false);
    bool result = isReachableNode(nodes, edges, src, des, vis);

    cout << "The given two nodes are Reachable or Not -->" << (result ? "true" : "false") << endl;

    return 0;
}
